synapse/storage/schema/main/delta/73/22_un_partial_stated_event_stream.sql

```diff
@@ -0,0 +1,34 @@
+/* Copyright 2022 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Stream for notifying that an event has become un-partial-stated.
+CREATE TABLE un_partial_stated_event_stream(
+    -- Position in the stream
+    stream_id BIGINT PRIMARY KEY NOT NULL,
+
+    -- Which instance wrote this entry.
+    instance_name TEXT NOT NULL,
+
+    -- Which event has been un-partial-stated.
+    event_id TEXT NOT NULL REFERENCES events(event_id) ON DELETE CASCADE,
+
+    -- true iff the `rejected` status of the event changed when it became
+    -- un-partial-stated.
+    rejection_status_changed BOOLEAN NOT NULL
+);
+
+-- We want an index here because of the foreign key constraint:
+-- upon deleting an event, the database needs to be able to check here.
+CREATE UNIQUE INDEX un_partial_stated_event_stream_room_id ON un_partial_stated_event_stream (event_id);
```

Comment:
I guess I should be a bit more stringent about researching this (possibly need to 'prove' that we only make an event partial-stated when it's initially persisted(?)), but I can't for the life of me imagine how it makes sense to turn a full-state event into a partial-stated event (that we then intend to make full-state again).
When a user leaves a room, we should keep the events and state around so that the user can access the room in their 'archive' section.
Only when a room is purged (e.g. because all users *forgot* the room and we do it automatically, or by admin action) should we start removing information we already hold. In this case, events will be removed but then so too will the stream entries for those events.

Reply:
I think I have confirmed this: an event can only be made partial-stated when it is created and possibly when it is de-outliered. But note that an event can't be both an outlier and partial-stated, so exactly one of those can be true for the lifetime of an event.
An event is only un-partial-stated upon deletion (at which point we also remove the stream row for it) or after we fetch the full state for it (it can't revert to being an outlier from this state, except by being deleted first and being persisted anew later).